### ISSUE TITLE
Detailed Training Configurations

### ISSUE THREAD

**Issue by hansongfang** [microsoft/TRELLIS#52]
The work is awesome and thank you for open-source!  I’m interested in training the model from scratch and could you please provide some instructions on the following:

Training Configurations:
- What optimizer and hyper parameters (learning rate, batch size, etc.) were used?
- How many steps were trained?  
- For Slat VAE-gs/raidanceField/flexicubes, how many views is used per object per training step? 

Thank you so much!

**Comment by JeffreyXiang** [comment]
Hi! 
- We use AdamW optimizer with `decay_rate=0.0` (equals Adam maybe?). The LR is set to constant `1e-4` and batch size is `256` for all the models.
- The VAE for GS/RF/Mesh is trained for 750K, 500K and 290K steps, respectively
- 1 view per object

**Comment by hansongfang** [comment]
Thank you for the information!

**Comment by lan-creator** [comment]
> Hi!
> 
> * We use AdamW optimizer with `decay_rate=0.0` (equals Adam maybe?). The LR is set to constant `1e-4` and batch size is `256` for all the models.
> * The VAE for GS/RF/Mesh is trained for 750K, 500K and 290K steps, respectively
> * 1 view per object

Hi, I noticed 256 batchsize for encoder and decoder training, but the code in https://github.com/microsoft/TRELLIS/blob/eeacb0bf6a7d25058232d746bef4e5e880b130ff/trellis/modules/sparse/attention/modules.py#L106-L107
which shows that there is no separate dim for batchsize of qkv.feat and qkv.coords. I also found that the shape of coords is always [N, 4] with [:, 0] set to 0. And swin attention here indicates only one batch can be trained at a time. So I wonder how to train decoder in 256 batch. Thank you for your reply.
https://github.com/microsoft/TRELLIS/blob/eeacb0bf6a7d25058232d746bef4e5e880b130ff/trellis/modules/sparse/attention/windowed_attn.py#L63-L85

**Comment by zoujx96** [comment]
> > Hi!
> > 
> > * We use AdamW optimizer with `decay_rate=0.0` (equals Adam maybe?). The LR is set to constant `1e-4` and batch size is `256` for all the models.
> > * The VAE for GS/RF/Mesh is trained for 750K, 500K and 290K steps, respectively
> > * 1 view per object
> 
> Hi, I noticed 256 batchsize for encoder and decoder training, but the code in
> 
> [TRELLIS/trellis/modules/sparse/attention/modules.py](https://github.com/microsoft/TRELLIS/blob/eeacb0bf6a7d25058232d746bef4e5e880b130ff/trellis/modules/sparse/attention/modules.py#L106-L107)
> 
> Lines 106 to 107 in [eeacb0b](/microsoft/TRELLIS/commit/eeacb0bf6a7d25058232d746bef4e5e880b130ff)
> 
>  if self._type == "self": 
>      qkv = self._linear(self.to_qkv, x) 
> 
> which shows that there is no separate dim for batchsize of qkv.feat and qkv.coords. I also found that the shape of coords is always [N, 4] with [:, 0] set to 0. And swin attention here indicates only one batch can be trained at a time. So I wonder how to train decoder in 256 batch. Thank you for your reply.
> [TRELLIS/trellis/modules/sparse/attention/windowed_attn.py](https://github.com/microsoft/TRELLIS/blob/eeacb0bf6a7d25058232d746bef4e5e880b130ff/trellis/modules/sparse/attention/windowed_attn.py#L63-L85)
> 
> Lines 63 to 85 in [eeacb0b](/microsoft/TRELLIS/commit/eeacb0bf6a7d25058232d746bef4e5e880b130ff)
> 
>  def sparse_windowed_scaled_dot_product_self_attention( 
>      qkv: SparseTensor, 
>      window_size: int, 
>      shift_window: Tuple[int, int, int] = (0, 0, 0) 
>  ) -> SparseTensor: 
>      """ 
>      Apply windowed scaled dot product self attention to a sparse tensor. 
>   
>      Args: 
>          qkv (SparseTensor): [N, *, 3, H, C] sparse tensor containing Qs, Ks, and Vs. 
>          window_size (int): The window size to use. 
>          shift_window (Tuple[int, int, int]): The shift of serialized coordinates. 
>          shift (int): The shift to use. 
>      """ 
>      assert len(qkv.shape) == 4 and qkv.shape[1] == 3, f"Invalid shape for qkv, got {qkv.shape}, expected [N, *, 3, H, C]" 
>   
>      serialization_spatial_cache_name = f'window_partition_{window_size}_{shift_window}' 
>      serialization_spatial_cache = qkv.get_spatial_cache(serialization_spatial_cache_name) 
>      if serialization_spatial_cache is None: 
>          fwd_indices, bwd_indices, seq_lens, seq_batch_indices = calc_window_partition(qkv, window_size, shift_window) 
>          qkv.register_spatial_cache(serialization_spatial_cache_name, (fwd_indices, bwd_indices, seq_lens, seq_batch_indices)) 
>      else: 
>          fwd_indices, bwd_indices, seq_lens, seq_batch_indices = serialization_spatial_cache

In [N, 4], the first dimension of the second dim is batch dimension. the model does support batch forward.

**Comment by HuynhNguyenPhuc** [comment]
> > > Hi!
> > > 
> > > * We use AdamW optimizer with `decay_rate=0.0` (equals Adam maybe?). The LR is set to constant `1e-4` and batch size is `256` for all the models.
> > > * The VAE for GS/RF/Mesh is trained for 750K, 500K and 290K steps, respectively
> > > * 1 view per object
> > 
> > 
> > Hi, I noticed 256 batchsize for encoder and decoder training, but the code in
> > [TRELLIS/trellis/modules/sparse/attention/modules.py](https://github.com/microsoft/TRELLIS/blob/eeacb0bf6a7d25058232d746bef4e5e880b130ff/trellis/modules/sparse/attention/modules.py#L106-L107)
> > Lines 106 to 107 in [eeacb0b](/microsoft/TRELLIS/commit/eeacb0bf6a7d25058232d746bef4e5e880b130ff)
> > if self._type == "self":
> > qkv = self._linear(self.to_qkv, x)
> > which shows that there is no separate dim for batchsize of qkv.feat and qkv.coords. I also found that the shape of coords is always [N, 4] with [:, 0] set to 0. And swin attention here indicates only one batch can be trained at a time. So I wonder how to train decoder in 256 batch. Thank you for your reply.
> > [TRELLIS/trellis/modules/sparse/attention/windowed_attn.py](https://github.com/microsoft/TRELLIS/blob/eeacb0bf6a7d25058232d746bef4e5e880b130ff/trellis/modules/sparse/attention/windowed_attn.py#L63-L85)
> > Lines 63 to 85 in [eeacb0b](/microsoft/TRELLIS/commit/eeacb0bf6a7d25058232d746bef4e5e880b130ff)
> > def sparse_windowed_scaled_dot_product_self_attention(
> > qkv: SparseTensor,
> > window_size: int,
> > shift_window: Tuple[int, int, int] = (0, 0, 0)
> > ) -> SparseTensor:
> > """
> > Apply windowed scaled dot product self attention to a sparse tensor.
> > ```
> >  Args: 
> >      qkv (SparseTensor): [N, *, 3, H, C] sparse tensor containing Qs, Ks, and Vs. 
> >      window_size (int): The window size to use. 
> >      shift_window (Tuple[int, int, int]): The shift of serialized coordinates. 
> >      shift (int): The shift to use. 
> >  """ 
> >  assert len(qkv.shape) == 4 and qkv.shape[1] == 3, f"Invalid shape for qkv, got {qkv.shape}, expected [N, *, 3, H, C]" 
> > 
> >  serialization_spatial_cache_name = f'window_partition_{window_size}_{shift_window}' 
> >  serialization_spatial_cache = qkv.get_spatial_cache(serialization_spatial_cache_name) 
> >  if serialization_spatial_cache is None: 
> >      fwd_indices, bwd_indices, seq_lens, seq_batch_indices = calc_window_partition(qkv, window_size, shift_window) 
> >      qkv.register_spatial_cache(serialization_spatial_cache_name, (fwd_indices, bwd_indices, seq_lens, seq_batch_indices)) 
> >  else: 
> >      fwd_indices, bwd_indices, seq_lens, seq_batch_indices = serialization_spatial_cache
> > ```
> 
> In [N, 4], the first dimension of the second dim is batch dimension. the model does support batch forward.

@zoujx96 I used dataset_toolkits and got the SLat which its coordinate shape is [N, 3]:
1. How to stack the list of SLat Sparse Tensor into [N, 4]. I had used sp.sparse_cat but it was still [N, 3]
2. Is all the models support batch forward, can I easily make it into a batch with [N, 4] and all other things the model will handle it?

**Comment by zoujx96** [comment]
@HuynhNguyenPhuc  After encoding and getting the latent, make sure to avoid doing `'coords': latent.coords[:, 1:]`. Yeah after batchifying model handle everything, you can verify by batch inference to see the rendering result.

**Comment by animemory** [comment]
> Hi!
> 
> * We use AdamW optimizer with `decay_rate=0.0` (equals Adam maybe?). The LR is set to constant `1e-4` and batch size is `256` for all the models.
> * The VAE for GS/RF/Mesh is trained for 750K, 500K and 290K steps, respectively
> * 1 view per object

Hi, are these different decoders using the same data?
